archive/miu_language/basic.lean

```diff
@@ -0,0 +1,249 @@
+/-
+Copyright (c) 2020 Gihan Marasingha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gihan Marasingha
+-/
+import tactic.linarith
+
+/-!
+# An MIU Decision Procedure in Lean
+
+The [MIU formal system](https://en.wikipedia.org/wiki/MU_puzzle) was introduced by Douglas
+Hofstadter in the first chapter of his 1979 book,
+[Gödel, Escher, Bach](https://en.wikipedia.org/wiki/G%C3%B6del,_Escher,_Bach).
+The system is defined by four rules of inference, one axiom, and an alphabet of three symbols:
+`M`, `I`, and `U`.
+
+Hofstadter's central question is: can the string "MU" be derived?
+
+It transpires that there is a simple decision procedure for this system. A string is derivable if
+and only if it starts with `"M"`, contains no other `"M"`s, and the number of `"I"`s in the string
+is congruent to 1 or 2 modulo 3.
+
+## The MIU System
+
+An _atom_ is any one of `M`, `I` or `U`. A _string_ is a finite sequence of zero or more symbols.
+To simplify notation, we write a sequence `[I,U,U,M]`, for example, as `IUUM`.
+
+The four rules of inference are:
+
+1. x`I` → x`IU`,
+2. `M`x → `M`xx,
+3. x`III`y → x`U`y,
+4. x`UU`y → xy,
+
+where the notation α → β is to be interpreted as 'if α is derivable, then β is derivable'.
+
+Additionally, we have an axiom
+
+* `MI` is derivable.
+
+In this file, the set of atoms and the set of derivable strings are represented as inductive types.
+-/
+
+
+namespace miu
+
+/-!
+### Basic data types
+-/
+
+/--
+Each MIU string consists of either an `M`, `I`, or `U`. Such an elementary unit is called an
+`miu_atom`. We represent `miu_atom` as an enumerated type.
+-/
+inductive miu_atom : Type
+| M : miu_atom
+| I : miu_atom
+| U : miu_atom
+
+open miu_atom 
+
+/--
+We show that the type `miu_atom` is inhabited, giving `M` (for no particular reason) as the default
+element.
+-/
+instance miu_atom_inhabited : inhabited miu_atom :=
+inhabited.mk M
+
+/--
+A simple function from `miu_atom` to `string`.
+-/
+def miu_atom.repr : miu_atom → string 
+| M := "M"
+| I := "I"
+| U := "U"
+
+/--
+A representation of an `miu_atom`.
+-/
+instance : has_repr miu_atom :=
+⟨λ u, u.repr⟩
+
+/--
+For simplicity, an `miustr` is just a list of `miu_atom`.
+-/
+def miustr := list miu_atom 
+
+/--
+We want to use list membership ...
+-/
+instance : has_mem miu_atom miustr :=
+  ⟨list.mem⟩
+
+/--
+... and list append.
+-/
+instance : has_append miustr :=
+⟨list.append⟩
```

Comment:
How about:

```suggestion
/--
For simplicity, an `miustr` is just a list of `miu_atom`.
-/
@[derive [has_mem, has_append]]
def miustr := list miu_atom 
```

(perhaps with an additional comment saying what `derive` does?)

Reply:
Thanks. I didn't know about @derive!

Reply:
Still not sure what to do here. I managed to get
```lean
@[derive has_append]
def miustr := list miu_atom
``` to work, but the same technique didn't work for classes, such as `has_mem` with more than one parameter.

Reply:
Ah, of course, sorry, I forgot about the other parameter of `has_mem`. Just leave that one as is (or use `by { dsimp [miustr], apply_instance }`).

Reply:
Thanks. I'll just leave as is.